packages/@aws-cdk/aws-apigatewayv2/README.md

```diff
@@ -19,6 +19,235 @@
+see [About WebSocket APIs in API Gateway](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api-overview.html) 
+and [HTTP APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html). 
+
+For more information about `REST APIs`, see [Working with REST APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-rest-api.html). To create [REST APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-rest-api.html) with AWS CDK, use `@aws-cdk/aws-apigateway` instead.
+
+
+### HTTP API
+
+
+HTTP APIs enable you to create RESTful APIs with lower latency and lower cost than REST APIs.
+
+You can use HTTP APIs to send requests to AWS Lambda functions or to any routable HTTP endpoint.
+
+For example, you can create an HTTP API that integrates with a Lambda function on the backend. When a client calls your API, 
+API Gateway sends the request to the Lambda function and returns the function's response to the client. 
+
+HTTP API supports both `Lambda Proxy Integration` and `HTTP Proxy Integration`. 
+See [Working with AWS Lambda Proxy Integrations for HTTP APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html) 
+and [Working with HTTP Proxy Integrations for HTTP APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-http.html) 
+for more information.
+
+
+
+Use `HttpApi` to create HTTP APIs with Lambda or HTTP proxy integration.
+
+
+```ts
+// Create a vanilla HTTP API with no integration targets
+const httpApi1 = new apigatewayv2.HttpApi(stack, 'HttpApi1');
+
+// Create a HTTP API with Lambda Proxy Integration as its $default route
+const httpApi2 = new apigatewayv2.HttpApi(stack, 'HttpApi2', {
+  targetHandler: handler
+});
+
+// Create a HTTP API with HTTP Proxy Integration as its $default route
+const httpApi3 = new apigatewayv2.HttpApi(stack, 'HttpApi3', {
+  targetUrl: checkIpUrl
+});
+```
+
+## Route
+
+`Routes` direct incoming API requests to backend resources. `Routes` consist of two parts: an HTTP method and a resource path—for example, 
+`GET /pets`. You can define specific HTTP methods for your route, or use the `ANY` method to match all methods that you haven't defined for a resource. 
+You can create a `$default route` that acts as a catch-all for requests that don’t match any other routes. See 
+[Working with Routes for HTTP APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-routes.html).
+
+When you create HTTP APIs with either `Lambda Proxy Integration` or `HTTP Proxy Integration`, the `$default route` will be created as well.
+
+
+
+To create a root route for an existing `Api` resource
+
+```ts
+// create a root route for the API
+httpApi.root = new apigatewayv2.Route(stack, 'RootRoute', {
+  api: httpApi,
+  httpPath: '/',
+  integration: integRootHandler
+});
+```
```

Comment:
1. By specifying route, it's already clear that the path is `'/'`, this should not be needed.

2. You have not yet mentioned anything so far in the document about integration but already have an `integration: integRootHandler`. The user is bound to get confused on what this is.

3. Can't we just pass a `rootIntegration` (or something along those lines) in the constructor and let the CDK set this route up correctly?

Reply:
OK. Let me implement this.

Reply:
Hi @nija-at 

I am considering to make it this way. What do you think?


```ts
// create a basic HTTP API with http proxy integration as the $default route
const api = new apigatewayv2.HttpApi(stack, 'HttpApi', {
  targetUrl: checkIpUrl,
});

// create a lambda proxy integration for this api 
const rootIntegration = new apigatewayv2.LambdaProxyIntegration(stack, 'RootIntegration', {
  api,
  targetHandler: rootHandler
});

// pass it to addRootRoute() to initialize the root route of this api
// HTTP ANY /
api.addRootRoute(rootIntegration)
  // HTTP GET /foo
  .addLambdaRoute('foo', 'Foo', {
    target: handler,
    method: apigatewayv2.HttpMethod.GET
  })
  // HTTP ANY /foo/aws
  .addHttpRoute('aws',  'AwsPage', {
    targetUrl: awsUrl,
    method: apigatewayv2.HttpMethod.ANY
  })
  // HTTP ANY /foo/aws/checkip
  .addHttpRoute('checkip', 'CheckIp', {
    targetUrl: checkIpUrl,
    method: apigatewayv2.HttpMethod.ANY
  });
```

Reply:
Can you describe what you're trying to do in this code? I'm not able to follow on how the final HTTP API will be configured when this code is executed. 

Perhaps you can start by specifying what end state you are trying to achieve with your HTTP API code snippet (which routes should go to which proxy) and we can work backwards on what the API design for this would look like.

Reply:
Let me take a stab at this at a high level, hoping that'll help move this forward -

```ts
const api = new HttpApi(stack, 'HttpApi', {
  defaultIntegration: new HttpProxyIntegration(..., {
    url: 'http://my-default-endpoint.app/'
  }),
});

api.addRoute('/', {
  methods: [ Http.GET, Http.POST ]
  integration: new HttpProxyIntegration(..., {
    url: 'http://my-root-endpoint.app/'
  }),
});

const getbooksHandler = new lambda.Function(...);
const getbookReviewsHandler = new lambda.Function(...);

api.addRoute('/books', {
  methods: [ Http.GET ],
  integration: new LambdaProxyIntegration(..., {
    handler: getbooksHandler,
    ...
  }),
}

api.addRoute('/books/reviews', {
  methods: [ Http.GET ],
  integration: new LambdaProxyIntegration(..., {
    handler: getbookReviewsHandler
  }),
});
```

This piece of code sets the following routes -
* `GET /books` → lambda function `getbooksHandler`
* `GET /books/reviews` → lambda function `getbookReviewsHandler
* `GET /` → http endpoint `GET http://my-root-endpoint.app/`
* `POST /` → http endpoint `POST http://my-root-endpoint.app/`
* `POST /` → http endpoint `POST http://my-root-endpoint.app/`
* `PUT /` → http endpoint `PUT http://my-default-endpoint.app/`
* `PUT /books/reviews` → http endpoint `PUT http://my-default-endpoint.app/books/reviews`

This is not a full working code, but I thought I'll take a stab at one way this could be modeled. 
Does that help? How does that look to you?

Reply:
Thanks @nija-at  this is very helpful.

Let me follow your high level example above and update it.  The only concern to me is that each `Route` resource can only have one method, which will build its required [RouteKey](https://docs.aws.amazon.com/zh_tw/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-route.html#cfn-apigatewayv2-route-routekey) property. The RouteKey is in **'{method} {path}'** format. So in this example

```ts
api.addRoute('/', {
  methods: [ Http.GET, Http.POST ]
  integration: new HttpProxyIntegration(..., {
    url: 'http://my-root-endpoint.app/'
  }),
});
```
It will not be satisfied in a single `Route` resource and will require two Routes, one with `GET /` as the RouteKey and the other `POST /`.  Of course we can create them in `addRoute()` under the hood but it might be kind of confusing as `addRoute()` actually would generate more than one Routes.

What do you think?

Reply:
If we change the method name to `addRoutes`, would that make it clearer?

Reply:
Hi @nija-at 

Now it supports `addRoutes()` as you advise. Please check the following working code:

```ts
// create a basic HTTP API with http proxy integration as the $default route
const api = new apigatewayv2.HttpApi(stack, 'HttpApi', {
  targetUrl: defaultUrl,
});

api.addRoutes('/', 'RootRoute', {
  methods: [HttpMethod.GET, HttpMethod.POST],
  integration: new HttpProxyIntegration(stack, 'RootInteg', {
    api,
    targetUrl: rootUrl
  })
});

api.addRoutes('/books', 'GetBooksRoute', {
  methods: [HttpMethod.GET],
  integration: new LambdaProxyIntegration(stack, 'getbooksInteg', {
    api,
    targetHandler: getbooksHandler,
  }),
});

api.addRoutes('/books/reviews', 'GetBookReviewRoute', {
  methods: [HttpMethod.GET],
  integration: new LambdaProxyIntegration(stack, 'getBookReviewInteg', {
    api,
    targetHandler: getbookReviewsHandler
  })
});
```


Do we still need `addRootRoute()`, `addLambdaRoute()` and `addHttpRoute()` as I previously proposed? They all return single `Route` resource and is very handy to build the Routes top down from root(`/`), e.g.

```ts
// create a basic HTTP API with http proxy integration as the $default route
const api = new apigatewayv2.HttpApi(stack, 'HttpApi', {
  targetUrl: defaultCatchAllUrl,
});

// create a lambda proxy integration for this api 
const rootIntegration = new apigatewayv2.LambdaProxyIntegration(stack, 'RootIntegration', {
  api,
  targetHandler: rootHandler
});

// pass it to addRootRoute() to initialize the root route of this api
// HTTP ANY /
const root = api.addRootRoute(rootIntegration)

// HTTP GET /books
const books = root.addLambdaRoute('books', 'GetBooks', {
    target: getbooksHandler,
    method: apigatewayv2.HttpMethod.GET
})
// HTTP ANY /books/reviews
const reviews = books.addLambdaRoute('reviews',  'BooksReviews', {
    target: getbookReviewsHandler,
    method: apigatewayv2.HttpMethod.ANY
})
```

Reply:
> const api = new apigatewayv2.HttpApi(stack, 'HttpApi', {
  targetUrl: defaultUrl,
});

This `targetUrl` still feels a bit awkward. Why not follow the same pattern as done with routes using the `integration` property, but instead call it `defaultIntegration`? See my example above. 
I like the `LambdaProxyIntegration` and `HttpProxyIntegration`. They act as nice extensibility points when new features need to be included. Having just one string property (`targetUrl`) is not a very useful extensibility point.

> api.addRoutes('/', 'RootRoute', {
  methods: [HttpMethod.GET, HttpMethod.POST],
  integration: new HttpProxyIntegration(stack, 'RootInteg', {
    api,
    targetUrl: rootUrl
  })
});

The user should not have to pass `api` into `HttpProxyIntegration`. This should automatically be computed because of the call `api.addRoutes()`. This is for two reason - (a) makes the API cleaner, and (b) possibility of human error where a different API is accidentally passed in here.

> Do we still need addRootRoute(), addLambdaRoute() and addHttpRoute() as I previously proposed?

Sure. We can have these, but their APIs need to be designed similar to the other one. Looking at your examples, I'm not entirely sure we have the right API design for these yet. 

Trying to add so many things is only going to slow us down in the PR. I would instead focus on getting one set of implementations in. We can always add these on a later PR. 

Hope that helps.

Reply:
Hi @nija-at 


> The user should not have to pass `api` into `HttpProxyIntegration`. This should automatically be computed because of the call `api.addRoutes()`. This is for two reason - (a) makes the API cleaner, and (b) possibility of human error where a different API is accidentally passed in here.

Do you expect this

```ts
api.addRoutes('/', 'RootRoute', {
  methods: [HttpMethod.GET, HttpMethod.POST],
  integration: new HttpProxyIntegration(stack, 'RootInteg', {
    api,
    targetUrl: rootUrl
  })
});
```

to be something like this?

```ts
api.addRoutes('/', 'RootRoute', {
  methods: [HttpMethod.GET, HttpMethod.POST],
  integration: new HttpProxyIntegration(stack, 'RootInteg', {
    targetUrl: rootUrl
  })
});
```
I agree with you but the `HttpProxyIntegration` implements `Integration` resource which requires the `Api` property. I have no idea how to create it in api.addRoutes() without specifying the `api`. Would be very appreciated if you could modify my code to help me move a little bit forward.

>Trying to add so many things is only going to slow us down in the PR. I would instead focus on getting one set of implementations in. We can always add these on a later PR.

Sure. Let me remove them and keep this PR clean and simpler.

Reply:
> Would be very appreciated if you could modify my code to help me move a little bit forward

We have several instances of this pattern in the CDK code base. We call it the bind pattern. Take a look at lambda-event-sources and how the `bind()` method on these work.

Reply:
BTW, let me explain why we still need `HttpProxyIntegration` and `LambdaProxyIntegration` in addition to the `Integratoin` resource.

It's still very complex to construct a Lambda proxy integration or Http proxy integration simply with the `Integration` class as we have to construct  and specify the `integrationUri` for Lambda like [this](https://github.com/aws/aws-cdk/blob/90204bb897a7840c882ede043ae347767e9505bd/packages/%40aws-cdk/aws-apigatewayv2/lib/integration.ts#L168) and make sure we  specify correct `integrationType` for both( `AWS_PROXY` for `LambdaProxyIntegration` while `HTTP_PROXY` for  `HttpProxyIntegration`). If we are building a http proxy integration all we must specify should be the HTTP target URL and keep others as as default. And in lambda proxy integration we just require the handler function while leaving other options remain the default.